contracts/src/v0.8/PermissionedForwardProxy.sol

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./ConfirmedOwner.sol";
+import {getRevertMsg} from "./utils/utils.sol";
+
+/**
+ * @title PermissionedForwardProxy
+ * @notice This proxy is used to forward calls from sender to target. It maintains
+ * a permission list to check which sender is allowed to call which target
+ */
+contract PermissionedForwardProxy is ConfirmedOwner {
+  mapping(address => address) public forwardPermissionList;
+
+  constructor() ConfirmedOwner(msg.sender) {}
+
+  /**
+   * @notice Verifies if msg.sender has permission to forward to target address and then forwards the handler
+   * @param target address of the contract to forward the handler to
+   * @param handler bytes to be passed to target in call data
+   */
+  function forward(address target, bytes calldata handler) external {
+    require(forwardPermissionList[msg.sender] == target, "Forwarding permission not found");
+    (bool success, bytes memory payload) = target.call(handler);
+    if (!success) {
+      revert(getRevertMsg(payload));
+    }
+  }
+
+  /**
+   * @notice Adds permission for sender to forward calls to target via this proxy.
+   * Note that it allows to overwrite an existing permission
+   * @param sender The address who will use this proxy to forward calls
+   * @param target The address where sender will be allowed to forward calls
+   */
+  function addPermission(address sender, address target) external onlyOwner {
```

Comment:
usually more common to do `setPermission(sender, target, bool)` instead of having both `add` and `remove` functions.

Reply:
setPermission(sender, target, bool) will probably be confusing if we keep the address => address mapping (Removal doesn't need a target address). I'll keep this as two separate functions for now.